icu4c/source/common/lsr.cpp

```diff
@@ -0,0 +1,118 @@
+// © 2019 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+
+// lsr.cpp
+// created: 2019may08 Markus W. Scherer
+
+#include "unicode/utypes.h"
+#include "cmemory.h"
+#include "cstring.h"
+#include "lsr.h"
+
+U_NAMESPACE_BEGIN
+
+LSR::LSR(char prefix, const char *lang, const char *scr, const char *r, UErrorCode &errorCode) :
+        language(nullptr), script(nullptr), region(r), owned(nullptr),
+        languageLength(0), scriptLength(0), regionIndex(indexForRegion(region)) {
+    if (U_SUCCESS(errorCode)) {
+        // Number of bytes for language + script with 2 prefixes and NUL terminators.
+        int32_t langSize = uprv_strlen(lang) + 1, scriptSize = uprv_strlen(scr) + 1;
+        int32_t langScriptSize = 2 + langSize + scriptSize;
+        char *p = owned = (char *)uprv_malloc(langScriptSize);
+        if (owned != nullptr) {
+            language = p;
+            *p++ = prefix;
+            uprv_memcpy(p, lang, langSize);
+            script = p += langSize;
+            *p++ = prefix;
+            uprv_memcpy(p, scr, scriptSize);
+            languageLength = langSize;
+            scriptLength = scriptSize;
+        } else {
+            errorCode = U_MEMORY_ALLOCATION_ERROR;
+        }
+    }
+}
+
+LSR::LSR(LSR &&other) U_NOEXCEPT :
+        language(other.language), script(other.script), region(other.region), owned(other.owned),
+        languageLength(other.languageLength), scriptLength(other.scriptLength),
+        regionIndex(other.regionIndex) {
+    if (owned != nullptr) {
+        other.language = other.script = "";
+        other.owned = nullptr;
+        other.languageLength = other.scriptLength = 0;
+    }
+}
+
+LSR::~LSR() {
+    uprv_free(owned);
+}
+
+LSR &LSR::operator=(LSR &&other) U_NOEXCEPT {
+    language = other.language;
+    script = other.script;
+    region = other.region;
+    languageLength = other.languageLength;
+    scriptLength = other.scriptLength;
+    regionIndex = other.regionIndex;
+    delete owned;
+    owned = other.owned;
+    if (owned != nullptr) {
+        other.language = other.script = "";
+        other.owned = nullptr;
+        other.languageLength = other.scriptLength = 0;
+    }
+    return *this;
+}
+
+UBool LSR::operator==(const LSR &other) const {
+    return languageLength == other.languageLength &&
+        uprv_memcmp(language, other.language, languageLength) == 0 &&
```

Comment:
memcmp() is faster (can look at several bytes at a time, rather than never look past a NUL byte) so if we have the length that's better. also, if the strings share a long common prefix but have different lengths, strcmp() wastes a lot of time while the length comparison bails out immediately.

however, these strings are super short and i am removing the lengths, so i am going back to strcmp().